dropshot/src/websocket.rs

```diff
@@ -0,0 +1,410 @@
+// Copyright 2022 Oxide Computer Company
+/*!
+ * Implements websocket upgrades as an Extractor for use in API route handler
+ * parameters to indicate that the given endpoint is meant to be upgraded to
+ * a websocket.
+ *
+ * This exposes a [`HyperUpgraded`] to a user-provided async future,
+ * which will be spawned to handle the incoming connection.
+ */
+
+use crate::api_description::ExtensionMode;
+use crate::{
+    ApiEndpointBodyContentType, Extractor, ExtractorMetadata, HttpError,
+    RequestContext, ServerContext,
+};
+use async_trait::async_trait;
+use http::{header, Response, StatusCode};
+use hyper::upgrade::OnUpgrade;
+use hyper::{upgrade, Body};
+use schemars::JsonSchema;
+use serde_json::json;
+use sha1::{Digest, Sha1};
+use slog::Logger;
+use std::future::Future;
+use std::sync::Arc;
+use tokio::task::JoinHandle;
+
+// re-export such that consumers need not version-match with us in their Cargo.toml
+pub use hyper::upgrade::Upgraded as HyperUpgraded;
+
+/**
+ * WebsocketUpgrade is an Extractor used to upgrade and handle an HTTP request
+ * as a websocket when present in a Dropshot endpoint's function arguments.
+ *
+ * The consumer of this *must* call [WebsocketUpgrade::handle] (or if they do
+ * not wish to upgrade a given connection, [WebsocketUpgrade::do_not_upgrade])
+ * on the owned instance in the endpoint's fn, or else WebsocketUpgrade's
+ * implementation of [Drop::drop] will panic.
+ */
+#[derive(Debug)]
+pub struct WebsocketUpgrade(Option<WebsocketUpgradeInner>);
+
+/**
```

Comment:
Did you mean to put some docs here and on `WebsocketUpgradeResponse`?

Reply:
:woman_facepalming: sure did! thanks